### PR TITLE
[wip] Add failing test case to spot an accountsDb corruption bug

### PR DESCRIPTION
**Pull request by ryoqun**
Related to #7117.

If this is a real bug, this is scary. Basically, this corrupts the very integrity of AccountsDB.

The problem here is that a StorageEntry with a zero-lamport account data is prematurely/incorrectly flushed, and **the account becomes non-zero balance again, namely the older version of account is restored**, resulting in unexpected internal hash error.

(I've been concerned with these problems in my mind since the first draft of #7013, which resorted to bizarre solution because of this, this indeed seems to be a real issue, now)

### CODE REVIEW

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -1735,6 +1735,86 @@ pub mod tests {
...
+
+        let accounts = AccountsDB::new_single();
+
+        let mut current_slot = 1;
+        accounts.store(current_slot, &[(&pubkey, &account)]);
+        accounts.add_root(current_slot);
+
+        current_slot += 1;
+        accounts.store(current_slot, &[(&pubkey, &zero_lamport_account)]);
+        for _ in 0..33000 {
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs]: If I reduce this number, this test passes...

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -1735,6 +1735,86 @@ pub mod tests {
...
+        accounts.add_root(current_slot);
+
+        error!("doesn't fail:");
+        assert_load_account(&accounts, current_slot, pubkey, zero_lamport);
+
+        purge_zero_lamport_accounts(&accounts, current_slot);
+        let accounts = reconstruct_accounts_db_via_serialization(accounts, current_slot);
+
+        error!("does fail due to a reconstruction bug:");
+        assert_load_account(&accounts, current_slot, pubkey, zero_lamport);
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs]: This should behave exactly same as [this assertion](https://github.com/solana-labs/solana/pull/7143/files#diff-2099c5256db4eb5975c8834af38f6456R1809).

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -1735,6 +1735,86 @@ pub mod tests {
...
+        accounts.store(current_slot, &[(&pubkey, &zero_lamport_account)]);
+        for _ in 0..33000 {
+            accounts.store(current_slot, &[(&filler_account_pubkey, &filler_account)]);
+        }
+        accounts.add_root(current_slot);
+
+        error!("doesn't fail:");
+        assert_load_account(&accounts, current_slot, pubkey, zero_lamport);
+
+        purge_zero_lamport_accounts(&accounts, current_slot);
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs]: If I remove this line, this test passes.

File: runtime/src/accounts_db.rs (hunk 2 above)
Comment (ryoqun) [reply runtime/src/accounts_db.rs]: This fails like [this](https://buildkite.com/solana-labs/solana/builds/15683#9428b65b-a60e-4318-97a6-336d5c87e3bf/79-2602).

**Comment by mvines** [comment]
(@ryoqun - just a minor suggestion: it would be helpful if this PR title was a bit more descriptive.  "Add failing test case to spot a bug": what bug?   Perhaps "Add failing test case to spot an accountsDb corruption bug")

**Comment by ryoqun** [comment]
> (@ryoqun - just a minor suggestion: it would be helpful if this PR title was a bit more descriptive. "Add failing test case to spot a bug": what bug? Perhaps "Add failing test case to spot an accountsDb corruption bug")

Exactly agreeable. :) Rename done.

**Comment by ryoqun** [comment]
Closing as this is merged as part of another PR: https://github.com/solana-labs/solana/pull/7190/files#diff-2099c5256db4eb5975c8834af38f6456R1944

**Pull request closed without merging**